docs/proposals/resource_management.md

```diff
@@ -0,0 +1,395 @@
+ - If a container exits, all its anonymous pages (tmpfs) get accounted to the machine (root). With pod level cgroups, that usage can be attributed to the pod.
+ - Let containers share resources - with pod level limits, containers can choose to burst on demand and still be capped at their pod limits.
+
+## Design
+
+High level requirements for the design are as follows:
+ - Do not break existing users. Ideally, there should be no changes to the Kubernetes API semantics.
+ - Support multiple cgroup managers - systemd, cgroupfs, etc.
+
+How we intend to achieve these high level goals is covered in greater detail in the Implementation Plan.
+
+Examples in the next section.
+
+### Proposed cgroup hierarchy:
+
+For the initial implementation we will only support limits for cpu and memory resources.
+
+#### Quality of Service classes
+
+A pod can belong to one of the following 3 QoS classes: Guaranteed, Burstable, and Best-Effort, in decreasing order of priority.
+We use the following denotation:
+
+`G⇒ Guaranteed QoS, Bu⇒ Burstable QoS, BE⇒ Besteffort Qos`
+
+#### Guaranteed
+
+Guaranteed pods will be placed at the `Root` cgroup by default. We would have a flag `guaranteed-qos-at-root` which would allow the user to explicitly specify if he needs a separate `G` cgroup for the Guaranteed Qos class. So by default the flag is set to true.
+
+```
+/PodUID/cpu.quota = cpu limit of Pod  
+/PodUID/cpu.shares = cpu request of Pod  
+/PodUID/m.limit = memory limit of Pod
+```
+
+`Note: In case guaranteed-qos-at-root flag is used to enable a separate Guaranteed class cgroup, we then create and set Cpu and Memory limits on the G cgroup aswell`
+
+Example:
+We have two pods Pod1 and Pod2 having Pod Spec given below
+
+```yaml
+kind: Pod
+metadata:
+    name: Pod1
+spec:
+    containers:
+        name: foo
+            resources:
+                limits:
+                    cpu: 10m
+                    memory: 1Gi
+        name: bar
+            resources:
+                limits:
+                    cpu: 100m
+                    memory: 2Gi
+```
+
+```yaml
+kind: Pod
+metadata:
+    name: Pod2
+spec:
+    containers:
+        name: foo
+            resources:
+                limits:
+                    cpu: 20m
+                    memory: 2Gii
+```
+
+Pod1 and Pod2 are both classified as Guaranteed and are nested under the `Root` cgroup.
+
+```
+/Pod1/cpu.quota = 110m  
+/Pod1/cpu.shares = 110m  
+/Pod2/cpu.quota = 20m  
+/Pod2/cpu.shares = 20m  
+/Pod1/m.limit = 3Gi  
+/Pod2/m.limit = 2Gi
+```
+
+#### Burstable
+
+We have the following resource parameters for the `Bu` cgroup.
+
+```
+/Bu/cpu.shares = summation of cpu requests of all Bu pods  
+/Bu/PodUID/cpu.quota = Pod Cpu Limit  
```

Comment:
Seems like to maximize utilization, all burstable pods should have the same quota. This setting should not impact the guaranteed class' access to the CPU. If that is true, would it make sense to set only aggregate quota on the /Bu cgroup instead? Or, are there other goals of the burstable limit besides protecting the guaranteed class?